# Express framework
- error handling


## 5-1. error handling
    - 에러 처리는 Express가 어떻게 동기적으로, 비동기적으로 모두 발생하는 에러를 잡아내고 처리하는지와 관려되어 있습니다.
    Express는 기본 에러 처리 함수를 가지고 있기에 사용자는 직접 에러 처리 함수를 작성할 필요가 없습니다.
    
    
    <에러 잡아내기>
    - Express가 경로 핸들러 함수와 미들웨어를 실행하는 동안 발생하는 모든 에러를 잡아내는 매우 중요합니다.
    - 라우트 핸들러 함수와 미들웨어 내부에 있는 동기적 코드에서 발생하는 에러들은 추가적 작업을 필요로 하지 않습니다.
    만약 동기적 코드가 에러를 던져내면, Express가 알아서 에러를 잡아내고 처리 할 것입니다. 예를 들어,
    
~~~Java Script
app.get('/', (request, response) {
  throw new Error('BROKEN') // Express 가 이 에러를 스스로 잡아냅니다.
  }) 
~~~

    - 라우트 핸들러 함수와 미들웨어에 의해 시작된 비동기적 함수로부터 반환된 에러들은, 사용자가 반드시 next()함수로 에러들을 넘겨야 합니다. 
    - next()함수는 Express가 에러들을 잡아내고, 처리하는 곳입니다. 예를 들어,
    
~~~Java Script
app.get('/', (request, response, next) =>{
  fs.readfile('/file-does-not-exist', (error, data) =>{
    if(error){
      next(error);// 에러들을 Express로 넘겨야 합니다.
    }else{
      response.send(data);
     }
   })
  })
  ~~~
  
  ## 5-2. default error handler
  
    - Express는 앱에서 마주할 수 있는 어느 에러들이든지 관리하는 내장형 에러 핸들러를 가지고 있습니다. 
     이 기본 에러-핸들링 미들웨어 함수는 미들웨어 함수 스택의 끝에 추가됩니다.
    - 만약 사용자가 에러를 next()함수로 넘기고, 사용자가 직접 작성한 에러 핸들러로 에러를 처리하지 않는다면, 에러는 기본 내장형 에러핸들러로 
      처리 될 것입니다. 
    - 에러는 스택 트레이스와 함께 사용자에게 보여질것입니다.. 스택 트레이스는 생산 환경에 포함되지 않습니다.
    - 생산성 모드에서 애플리케이션을 실행하기 위해, 환경 변수 NODE_ENV를 생산성으로 설정하세요. 
    - 에러가 사용자에게 보여질 때 아래의 정보가 response에 추가 될 것입니다.

    - response.statusCode는 error.status(혹은 error.statusCode)로 부터 설정됩니다. 만약 이 값이 4** 혹은 5** 범위의 밖이라면 500으로
     설정될 것입니다. 
    - response.statusMessage는 status code에 따라 설정됩니다. 
    - 바디 부분은 생산 환경에 있을때의 status code 메시지의 HTML 이며, 아니면 error.stack 일것입니다.
    - 어떤 헤더든지 error.headers 객체에 명세화 됩니다.
    - 만약 사용자가 적은 코드로 요청에 대한 응답을 시작한 후, next()함수를 에러와 함께 호출하면
      (예를 들어, 클라이언트에게 응답을 스트리밍하는 동안 에러를 만났을때),  Express 기본실행 에러 핸들러는 연결을 끊고, 
      요청을 처리하는데 있어 실패 할 것입니다. 
    - 따라서, 사용자가 직접 작성한 에러핸들러를 코드에 추가할 때, 사용자는 반드시 Express 내장형 기본 에러 핸들러 함수에게 처리를 
      맡겨야 합니다.헤더가 클라이언트에게 이미 보내어졌을 때는: 
    
~~~Java Script
function errorHandler ( error, request, response, next) {
if(response.headerSent){
return next(error);
}
response.status(500);
response.render('error', { error: err })
}
~~~
- NOTE: 당신 코드 내의 next()함수가 에러와 함께 한번 이상 호출된다면, 사용자가 작성한 에러 핸들링 미들웨어가 필요한 곳에 존재한다고
 하더라도 기본실행 에러 핸들러 함수가 실행 될 것입니다.



## Writing error handlers
   
   - Express는 앱에서 마주할 수 있는 어느 에러들이든지 관리하는 내장형 에러 핸들러를 가지고 있습니다. 
     이 기본 에러-핸들링 미들웨어 함수는 미들웨어 함수 스택의 끝에 추가됩니다.
     
     efine error-handling middleware functions in the same way as other middleware functions, except error-handling functions have four arguments instead of three: (err, req, res, next). For example:

app.use((err, req, res, next) => {
  console.error(err.stack)
  res.status(500).send('Something broke!')
})
You define error-handling middleware last, after other app.use() and routes calls; for example:

const bodyParser = require('body-parser')
const methodOverride = require('method-override')

app.use(bodyParser.urlencoded({
  extended: true
}))
app.use(bodyParser.json())
app.use(methodOverride())
app.use((err, req, res, next) => {
  // logic
})
Responses from within a middleware function can be in any format, such as an HTML error page, a simple message, or a JSON string.

For organizational (and higher-level framework) purposes, you can define several error-handling middleware functions, much as you would with regular middleware functions. For example, to define an error-handler for requests made by using XHR and those without:

const bodyParser = require('body-parser')
const methodOverride = require('method-override')

app.use(bodyParser.urlencoded({
  extended: true
}))
app.use(bodyParser.json())
app.use(methodOverride())
app.use(logErrors)
app.use(clientErrorHandler)
app.use(errorHandler)
In this example, the generic logErrors might write request and error information to stderr, for example:

function logErrors (err, req, res, next) {
  console.error(err.stack)
  next(err)
}
     
     
     
     
     
     
     
     
     
     
     
     
     

